### PR TITLE
altitude-profile component

### PR DESCRIPTION
**Pull request by norbertK**

### CODE REVIEW

File: src/modules/altitudeProfile/utils/altitudeProfileUtils.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+	gradientBg.addColorStop(xPoint, flatColor);
+	gradientBg.addColorStop(xPoint, steepColor);
+	return SlopeType.Steep;
+};
+
+export const startFlat = (gradientBg, xPoint) => {
+	gradientBg.addColorStop(xPoint, steepColor);
+	gradientBg.addColorStop(xPoint, flatColor);
+	return SlopeType.Flat;
+};
```
Comment (taulinger) [inline src/modules/altitudeProfile/utils/altitudeProfileUtils.js:20]: Avoid shallow functions (let's discuss that)

File: src/modules/utils/components/showCase/ShowCase.js [hunk 2]
```diff
@@ -20,12 +20,18 @@ import { closeBottomSheet, openBottomSheet } from '../../../../store/bottomSheet
...
-			= $injector.inject('CoordinateService', 'EnvironmentService', 'ShareService', 'UrlService', 'FileStorageService', 'ImportVectorDataService');
+		const { CoordinateService, EnvironmentService, ShareService, UrlService, FileStorageService, ImportVectorDataService } =
+			$injector.inject(
+				'CoordinateService',
+				'EnvironmentService',
+				'ShareService',
+				'UrlService',
+				'FileStorageService',
+				'ImportVectorDataService'
+			);
```
Comment (taulinger) [inline src/modules/utils/components/showCase/ShowCase.js:34]: Avoid unintended reformating of the code

File: src/modules/utils/components/showCase/ShowCase.js [hunk 3]
```diff
@@ -87,11 +91,14 @@ export class ShowCase extends BaElement {
 				}
 			};
 
-			const unsubscribe = observe(this._storeService.getStore(), state => state.modal, modal => resolveBeforeClosing(modal));
+			const unsubscribe = observe(
+				this._storeService.getStore(),
+				(state) => state.modal,
+				(modal) => resolveBeforeClosing(modal)
+			);
```
Comment (taulinger) [inline src/modules/utils/components/showCase/ShowCase.js:98]: Avoid unintended reformating of the code

File: src/modules/utils/components/showCase/ShowCase.js [hunk 4]
```diff
@@ -109,7 +116,11 @@ export class ShowCase extends BaElement {
 
 			// creates a PasswordCredentialPanel-element within a templateResult
 			const getCredentialPanel = () => {
-				return html`<ba-auth-password-credential-panel .url=${restrictedUrl} .authenticate=${authenticate} .onClose=${onClose}>`;
+				return html`<ba-auth-password-credential-panel
+					.url=${restrictedUrl}
+					.authenticate=${authenticate}
+					.onClose=${onClose}
+				></ba-auth-password-credential-panel>`;
```
Comment (taulinger) [inline src/modules/utils/components/showCase/ShowCase.js:123]: Avoid unintended reformating of the code

File: src/modules/utils/components/showCase/ShowCase.js [hunk 5]
```diff
@@ -153,47 +164,61 @@ export class ShowCase extends BaElement {
...
-					unsubscribe();
-				});
+				const unsubscribe = observe(
+					this._storeService.getStore(),
+					(state) => state.pointer.beingDragged,
+					() => {
+						closeBottomSheet();
+						unsubscribe();
+					}
+				);
```
Comment (taulinger) [inline src/modules/utils/components/showCase/ShowCase.js:180]: Avoid unintended reformating of the code

File: src/modules/utils/components/showCase/ShowCase.js [hunk 6]
```diff
@@ -153,47 +164,61 @@ export class ShowCase extends BaElement {
...
 				return html`<div>
 					<h3>Bottom Sheet autoclose with...</h3>
-					<div style="color: white;background-color: var(--warning-color);">observing store... </div>
-					<div style="color: white;background-color: var(--error-color);">i.e. dragging map</div>					
+					<div style="color: white;background-color: var(--warning-color);">observing store...</div>
+					<div style="color: white;background-color: var(--error-color);">i.e. dragging map</div>
```
Comment (taulinger) [inline src/modules/utils/components/showCase/ShowCase.js:184]: Avoid unintended reformating of the code

File: src/modules/utils/components/showCase/ShowCase.js [hunk 7]
```diff
@@ -153,47 +164,61 @@ export class ShowCase extends BaElement {
...
-							<div style="color: var(--text2);background-color: var(--secondary-bg-color);"><b>Value:</b>Lorem ipsum dolor sit amet, consetetur sadipscing elitr...</div>
-							<div style="display:flex"><ba-button .label=${'Wait & close'} @click=${onCloseAfterWait}></ba-button><ba-button .label=${'dismiss!'} @click=${onDismiss}></ba-button></div>
+							<div style="color: var(--text1);background-color: var(--secondary-color);"><b>ID:</b>Lorem ipsum dolor</div>
+							<div style="color: var(--text2);background-color: var(--secondary-bg-color);">
+								<b>Value:</b>Lorem ipsum dolor sit amet, consetetur sadipscing elitr...
+							</div>
+							<div style="display:flex">
+								<ba-button .label=${'Wait & close'} @click=${onCloseAfterWait}></ba-button
+								><ba-button .label=${'dismiss!'} @click=${onDismiss}></ba-button>
+							</div>
```
Comment (taulinger) [inline src/modules/utils/components/showCase/ShowCase.js:199]: Avoid unintended reformating of the code

File: src/modules/utils/components/showCase/ShowCase.js [hunk 8]
```diff
@@ -153,47 +164,61 @@ export class ShowCase extends BaElement {
...
 						</div>`;
 					case 2:
 						return html`<div>
 							<h3>Bottom Sheet ...</h3>
-							<div style="color: white;background-color: var(--warning-color);">waiting forever... </div>
-							<div style="color: white;background-color: var(--error-color);">until a new content for the bottom sheet comes</div>							
+							<div style="color: white;background-color: var(--warning-color);">waiting forever...</div>
+							<div style="color: white;background-color: var(--error-color);">
+								until a new content for the bottom sheet comes
+							</div>
```
Comment (taulinger) [inline src/modules/utils/components/showCase/ShowCase.js:207]: Avoid unintended reformating of the code

File: src/modules/utils/components/showCase/ShowCase.js [hunk 9]
```diff
@@ -153,47 +164,61 @@ export class ShowCase extends BaElement {
...
 			openBottomSheet(getContent(version));
 			version = nextVersion(version, 1, 3);
 		};
-		const menuitems = [{ label: 'Apple', icon: arrowUpSvg, action: () => emitNotification('Apple', LevelTypes.INFO) }, { label: 'Lemon', icon: arrowUpSvg, action: () => emitNotification('Lemon', LevelTypes.INFO) }, { label: 'Orange', action: () => emitNotification('Orange', LevelTypes.INFO) }, { label: 'Banana', icon: arrowUpSvg, disabled: true, action: () => emitNotification('Banana', LevelTypes.INFO) }];
+		const menuitems = [
+			{ label: 'Apple', icon: arrowUpSvg, action: () => emitNotification('Apple', LevelTypes.INFO) },
+			{ label: 'Lemon', icon: arrowUpSvg, action: () => emitNotification('Lemon', LevelTypes.INFO) },
+			{ label: 'Orange', action: () => emitNotification('Orange', LevelTypes.INFO) },
+			{ label: 'Banana', icon: arrowUpSvg, disabled: true, action: () => emitNotification('Banana', LevelTypes.INFO) }
+		];
```
Comment (taulinger) [inline src/modules/utils/components/showCase/ShowCase.js:221]: Avoid unintended reformating of the code

File: test/modules/altitudeProfile/utils/altitudeProfileUtils.test.js [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+$injector.registerSingleton('TranslationService', { translate: (key) => key });
+
+describe('Unit test functions altitudeProfileUtils.js', () => {
+	describe('check SlopeType', () => {
+		const slopeType = SlopeType.Flat;
+
+		it('check SlopeType length', () => {
+			expect(Object.keys(slopeType).length).toBe(4);
+		});
```
Comment (taulinger) [inline test/modules/altitudeProfile/utils/altitudeProfileUtils.test.js:12]: Tests the number of characters of "Flat", don`t know if it's the test you've wanted 😉

File: src/modules/altitudeProfile/components/AltitudeProfile.js [hunk 11]
```diff
@@ -0,0 +1,415 @@
...
+		this._initSurfaceTypes();
+		this._initAnotherTypeTypes();
+	}
+
+	/**
+	 * @override
+	 */
+	onInitialize() {
+		this.style.width = '100%';
+		this.style.height = '14em';
```
Comment (taulinger) [inline src/modules/altitudeProfile/components/AltitudeProfile.js:61]: We should have a test for the presence those important style properties

**Review by taulinger — COMMENTED** [review]
Yeah, pointing in the right direction 👍
The AltitudeProfile is structured well now, let's have a look at its remaining non-test-covered parts
and let's rework the `altitudeProfileUtils`.

File: src/modules/altitudeProfile/components/AltitudeProfile.js [hunk 12]
```diff
@@ -0,0 +1,427 @@
...
+
+		this.observe(
+			(state) => state.media.darkSchema,
+			(darkSchema) => this.signal(Update_Schema, darkSchema)
+		);
+		this.observe(
+			(state) => state.profile.coordinates,
```
Comment (taulinger) [inline src/modules/altitudeProfile/components/AltitudeProfile.js:72]: observed slice-of-state is `altitudeProfile` not `profile`

File: src/modules/altitudeProfile/components/AltitudeProfile.js [hunk 13]
```diff
@@ -0,0 +1,427 @@
...
+	static get FLAT_COLOR() {
+		return '#66eeff';
+	}
+
+	static get STEEP_COLOR() {
+		return '#ee4444';
+	}
+
+	static get tag() {
+		return 'ba-altitudeprofile-n';
```
Comment (taulinger) [inline src/modules/altitudeProfile/components/AltitudeProfile.js:425]: as already discussed the tag should be  `ba-altitude-profile`

**Review by taulinger — COMMENTED** [review]
Looks good for now, would suggest some minor layout optimizations to finish this task by  adding the following labels:

grafik

**i18n (de/en):**
DGM 25 / DHHN2016 / DGM 25 / DHHN2016
Höhe [m] / Eleavation [m]
Entfernung [m/km] / Distance [m/km] 

(Note: for distances >= 10,000m the unit currently (v3) changes to `km`.
Let's discuss this detail tomorrow...)

**Comment by norbertK** [comment]
Screenshot 2023-01-10 114852

**Comment by norbertK** [comment]
image

File: src/modules/altitudeProfile/components/AltitudeProfile.js [hunk 14]
```diff
@@ -0,0 +1,498 @@
...
+		return distUnit;
+	}
+
+	_getChartData(altitudeData, newDataLabels, newDataData) {
+		const _chartData = {
+			labels: newDataLabels,
+			datasets: [
+				{
+					data: newDataData,
+					label: 'Höhenprofil',
```
Comment (taulinger) [inline src/modules/altitudeProfile/components/AltitudeProfile.js:196]: Should be fetched from TranslationService

File: src/modules/altitudeProfile/components/AltitudeProfile.js [hunk 15]
```diff
@@ -0,0 +1,498 @@
...
+		return html`
+			<style>
+				${css}
+			</style>
+
+			<div class="chart-container" style="position: relative; height:100%; ">
+				<canvas class="altitudeprofile" id="route-altit_getAltitudeProfileude-chart"></canvas>
+
+				<div class="flex">
+					${translate('altitudeProfile_sumUp')}: ${sumUp} ${translate('altitudeProfile_sumDown')}: ${sumDown}
```
Comment (taulinger) [inline src/modules/altitudeProfile/components/AltitudeProfile.js:128]: Test for labels and values are missing

File: src/modules/altitudeProfile/components/AltitudeProfile.js [hunk 16]
```diff
@@ -0,0 +1,498 @@
...
+					x: { type: 'linear',
+						title: {
+							display: true,
+							text: translate('altitudeProfile_distance') + '[' + distUnit + ']'
+						}
+					},
+					y: { type: 'linear', beginAtZero: false,
+						title: {
+							display: true,
+							text: translate('altitudeProfile_alt') + '[m]'
```
Comment (taulinger) [inline src/modules/altitudeProfile/components/AltitudeProfile.js:374]: Test for `text` is missing

File: src/modules/altitudeProfile/components/AltitudeProfile.js [hunk 17]
```diff
@@ -0,0 +1,498 @@
...
+			options: {
+				responsive: true,
+				animation: false, // HINT: UX decision
+				maintainAspectRatio: false,
+
+				scales: {
+					x: { type: 'linear',
+						title: {
+							display: true,
+							text: translate('altitudeProfile_distance') + '[' + distUnit + ']'
```
Comment (taulinger) [inline src/modules/altitudeProfile/components/AltitudeProfile.js:368]: Test for `text` is missing

File: src/modules/altitudeProfile/components/AltitudeProfile.js [hunk 18]
```diff
@@ -0,0 +1,498 @@
...
+						grid: { drawOnChartArea: false }
+					}
+				},
+				events: ['mousemove', 'mousedown', 'mouseup', 'mouseout', 'click', 'touchstart', 'touchmove'],
+				plugins: {
+					title: {
+						align: 'end',
+						display: true,
+						text:
+							'DGM 25 / DHHN2016 / DGM 25 / DHHN2016'
```
Comment (taulinger) [inline src/modules/altitudeProfile/components/AltitudeProfile.js:390]: Should be retrieved from the TranslationService

File: src/modules/altitudeProfile/i18n/altitudeProfile.provider.js [hunk 19]
```diff
@@ -0,0 +1,30 @@
+export const provide = (lang) => {
+	switch (lang) {
+		case 'en':
+			return {
+				altitudeProfile_header: 'this is a header',
```
Comment (taulinger) [inline src/modules/altitudeProfile/i18n/altitudeProfile.provider.js:5]: Unused. Can we remove it?

**Review by taulinger — COMMENTED** [review]
See inline comments...

**Review by taulinger — APPROVED** [review]
Let's go!

**Pull request merged**